pallets/msa/src/lib.rs

```diff
@@ -1282,25 +1282,19 @@ impl<T: Config> MsaLookup for Pallet<T> {
 impl<T: Config> MsaValidator for Pallet<T> {
 	type AccountId = T::AccountId;
 
-	#[cfg(not(feature = "runtime-benchmarks"))]
-	fn ensure_valid_msa_key(key: &T::AccountId) -> Result<MessageSourceId, DispatchError> {
-		Self::ensure_valid_msa_key(key)
-	}
-
 	/// Since benchmarks are using regular runtime, we can not use mocking for this loosely bounded
 	/// pallet trait implementation. To be able to run benchmarks successfully for any other pallet
 	/// that has dependencies on this one, we would need to define msa accounts on those pallets'
 	/// benchmarks, but this will introduce direct dependencies between these pallets, which we
 	/// would like to avoid.
 	/// To successfully run benchmarks without adding dependencies between pallets we re-defined
 	/// this method to return a dummy account in case it does not exist
-	#[cfg(feature = "runtime-benchmarks")]
 	fn ensure_valid_msa_key(key: &T::AccountId) -> Result<MessageSourceId, DispatchError> {
 		let result = Self::ensure_valid_msa_key(key);
-		if result.is_err() {
+		if cfg!(feature = "runtime-benchmarks") && result.is_err() {
```

Comment:
As far as I know we are using mocks in tests and we are not calling these from regular tests. Now benchmarks are different in that regard that they are setup like a test but using the actual runtime instead of test runtime.
So in reality benchmarks run in two mode

1- as a test (when calling `cargo test` with runtime-benchmarks or all-features) 
2- as execution when called after build with runtime-benchmarks feature enabled

both of these modes are not using the same path of execution for these loosely coupled traits If I'm not mistaken.

We are doing this in benchmarks because we couldn't seed the db in a way that was required for benchmark, so we circumvent the whole seeding but if we can seed it correctly in benchmarks then we can get rid of these.

Reply:
the pattern is not posted

Reply:
Enddy's approach which adds seeding will also eliminate this problem

Reply:
Wow I can see it too , it was

Ifcfg(feature=="xyz) and else pattern

Reply:
Basically same function two branches